### PR TITLE
26 - Adding unetvit model

### LINKED ISSUE #26: Adding unetvit model
### 🚀 Feature: Refactor unetvit to train it with mobious dataset

* [x] [unetvit](https://github.com/mxochicale/unetvit/blob/main/src/unetvit/models/unetvit.py)
* [ ] Let's add details of an internal [shared path to debug and control version models](https://liveuclac-my.sharepoint.com/my?id=%2Fpersonal%2Fccaemxo%5Fucl%5Fac%5Fuk%2FDocuments%2Fnvidia%2Dclara%2Dagx%2Fready%2DSHARED%2DPATH%2Fmodels%2Dand%2Dvideos%2Funet%2Dvit&csf=1&web=1&e=7uNpQH&CID=eea3011a%2D44d1%2D4f94%2D9f5b%2D8b3ac00bacab&FolderCTID=0x012000E91CF1EF9659F94CB2FC662024FA4E44).
* [ ] [test_inference](https://github.com/mxochicale/unetvit/blob/main/tests/test_inference.py)
* [ ] Visualising training logs with wandb. See examples here https://github.com/xfetus/xfetus/blob/main/src/xfetus/models/dsrgan/train_baseline_ddpm.py https://github.com/xfetus/xfetus/blob/main/docs/figures/wandb-board.png

Main reference
* Chen, Jieneng, Yongyi Lu, Qihang Yu, Xiangde Luo, Ehsan Adeli, Yan Wang, Le Lu, Alan L. Yuille, and Yuyin Zhou. **"Transunet: Transformers make strong encoders for medical image segmentation."** arXiv preprint arXiv:2102.04306 (2021).
  * [google-citations](https://scholar.google.com/scholar?cites=1117067442557737826&as_sdt=2005&sciodt=0,5&hl=en)
  * [arxiv](https://arxiv.org/pdf/2102.04306v1)
  * [GitHub-project](https://github.com/Beckschen/TransUNet)
  * Videos: https://www.youtube.com/watch?v=jKBJITQ8xJY
  * **Implementation Details**: 
    * All Transformer backbones (i.e.,ViT) and ResNet-50 (denoted as “R-50”) were **pretrained** on ImageNet. 
    * Models are trained with SGD optimizer with learning rate 0.01, momentum 0.9 and weight decay 1e-4.
    * The default batch size is 24 and the default number of training iterations are 20k for ACDC dataset and 14k for Synapse dataset respectively.
    * The default batch size is 24 and the default number of training iterations are 20k for ACDC dataset and 14k for Synapse dataset respectively.
  * **Ablation study** : 
    * 1) the number of skip-connections (0-skip, 1-skip, 3-skip); 
      *  "We can see that adding more skip-connections generally leads to a better segmentation performance."
    * 2) input resolution; 
      * "As [4] indicated, increasing the effective sequence length shows robust improvements. For TransUNet, changing the resolution scale from 224×224 to 512×512 results in 6.88% improvement in average DSC, at the expense of a much larger computational cost."
    * 3) sequence length and patch size and 
      * "It is observed that a higher segmentation performance is usually obtained with smaller patch size. Note that the Transformer’s sequence length is inversely proportional to the square of the patch size"
    * 4) model scaling (Base DSC %77.48; Larger DSC %78.52)
      * “base” model, the hidden size D=12, number of layers=768, MLP size=3072, and number of heads=12
      * “large” model, the hidden size D=24, number of layers=1024, MLP size=4096, and number of heads=16

<details>
  <summary>See more references </summary>

* Drawing neural networks diagrams
   * Latex code for making neural networks diagrams: https://github.com/HarisIqbal88/PlotNeuralNet
   * Just trying https://inkscape.org/ (free and open-source vector graphics editor)
* Li, C., Wang, L. & Li, Y. Transformer and group parallel axial attention co-encoder for medical image segmentation. Sci Rep 12, 16117 (2022). https://doi.org/10.1038/s41598-022-20440-z
  * "we propose a new attention mechanism to highlight local foreground information, called group parallel axial attention (GPA). Furthermore, we effectively combine GPA with Transformer in encoder part of model. It can not only highlight the foreground information of samples, but also reduce the negative influence of background information on the segmentation results. Meanwhile, we introduced the sMLP block to improve the global modeling capability of network. Sparse connectivity and weight sharing are well achieved by applying it"
* HuBMAP | Pytorch | ViT for Segmentation [Train]: https://github.com/elcaiseri/Vision-Transformer-Segmentation/blob/main/hubmap-pytorch-vit-segmentation-train.ipynb, including Hyperparameters, Dataset, Visualize augmented data, Train Function, Creating Folds Column, The Real Training, Plot the results
* Transformers from Image Segmentation: https://www.youtube.com/watch?v=0SQrHVnji8I

</details>

### PR DESCRIPTION
**Pull request by mxochicale**
## Description
Fixes #26 Refactor scripts to train unetvit model with mobious dataset

## Type of change
Please delete options accordingly to the description.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

### Suggested Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation 
- [ ] My changes generate no new warnings 
- [ ] I have commented my code, particularly in hard-to-understand areas 
- [ ] I have read the [`CONTRIBUTING`](https://github.com/UCL/ready/blob/main/CONTRIBUTING.md) docs
- [ ] My code passes, following the style guidelines `pre-commit run -a` 
- [ ] My code is properly tested with `pytest -sv tests`
- [ ] I have requested review to this PR.
- [ ] I have addressed and marked as resolved all the review comments in my PR.
- [ ] Make sure your branch is up-to-date with main branch. See step 4 in [`CONTRIBUTING`](https://github.com/UCL/ready/blob/29-adds-contrib-code-conduct-pre-commit/CONTRIBUTING.md#pull-request-pr-and-merge-to-main-branch).
- [ ] Finally, I have selected `squash and merge`

### CODE REVIEW

**Review by mxochicale — COMMENTED** [review]
Thanks Tiantian, this is looking promising. I've left few comments
